test/workers/send_user_invitation_worker_test.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class SendUserInvitationWorkerTest < ActiveJob::TestCase
+  def test_provider_mailer_perform
+    Timecop.freeze(Time.zone.parse('2017-11-23 03:25:08 UTC +00:00')) do
+      ProviderInvitationMailer.any_instance.expects(:invitation).once
+
+      provider_account = FactoryBot.create(:provider_account)
+      invitation = FactoryBot.create(:invitation, account: provider_account)
+
+      assert_nil invitation.sent_at
+      SendUserInvitationWorker.new.perform(invitation.id)
+      assert invitation.reload.sent_at, Time.zone.now
+    end
+  end
+
+  def test_buyer_mailer_perform
+    Timecop.freeze(Time.zone.parse('2017-11-23 03:25:08 UTC +00:00')) do
+      InvitationMailer.any_instance.expects(:invitation).once
+
+      buyer_account = FactoryBot.create(:buyer_account)
+      invitation = FactoryBot.create(:invitation, account: buyer_account)
+
+      assert_nil invitation.sent_at
+      SendUserInvitationWorker.new.perform(invitation.id)
+      assert invitation.reload.sent_at
+    end
+  end
+
+  def test_handles_ssl_error
+    ProviderInvitationMailer.any_instance.expects(:invitation).raises(OpenSSL::SSL::SSLError)
+
+    invitation = FactoryBot.create(:invitation)
+
+    assert_nil invitation.sent_at
+    SendUserInvitationWorker.new.perform(invitation.id)
+    assert_nil invitation.reload.sent_at
+  end
+
+  def test_handles_smpt_error
+    ProviderInvitationMailer.any_instance.expects(:invitation).raises(Net::SMTPAuthenticationError)
+
+    invitation = FactoryBot.create(:invitation)
+
+    assert_nil invitation.sent_at
+    SendUserInvitationWorker.new.perform(invitation.id)
+    assert_nil invitation.reload.sent_at
+  end
+
+  def test_handles_socket_error
+    ProviderInvitationMailer.any_instance.expects(:invitation).raises(SocketError)
+
+    invitation = FactoryBot.create(:invitation)
+
+    assert_nil invitation.sent_at
+    SendUserInvitationWorker.new.perform(invitation.id)
+    assert_nil invitation.reload.sent_at
+  end
+
+  def test_handles_not_found_error
+    ProviderInvitationMailer.any_instance.expects(:invitation).raises(ActiveRecord::RecordNotFound)
+
+    invitation = FactoryBot.create(:invitation)
+
+    assert_nil invitation.sent_at
+    SendUserInvitationWorker.new.perform(invitation.id)
+    assert_nil invitation.reload.sent_at
+  end
+
+  # regression test for: https://github.com/3scale/system/pull/3316
+  def test_send_invitation_with_helper_tag
+    provider = FactoryBot.create(:simple_provider)
+    buyer = FactoryBot.create(:simple_buyer, provider_account: provider)
+    FactoryBot.create(:cms_email_template, system_name: 'invitation', provider: provider, published: '{% debug:help %}', rails_view_path: 'emails/invitation')
+    invitation = FactoryBot.create(:invitation)
+
+    ProviderInvitationMailer.any_instance.expects(:invitation)
```

Comment:
It doesn't mean the email will be rendered properly though. Not sure if it belongs to this test anyway.

Reply:
That isn't what the test is about, I think. It's asserting the mail is sent https://github.com/3scale/porta/pull/2092/files#diff-6667f7edb63c733c01adf25ebaef955aL154

Reply:
The test ensures that `ProviderInvitationMailer#invitation` is called, but keep in mind that stubbing that method with an expectation will make it return, in this case, `nil`, rather than actually _executing_ that method. Therefore, that CMS template created a couple lines before in the test won't ever get to the point of being rendered, thus making the test case pointless.

In order to test the rendering of the CMS template, with liquid tags being rendered and all, you'll need to get deeper than that, i.e. actually invoking `ProviderInvitationMailer#invitation` (not a stub), running it where it calls `ActionMailer::Base#mail` at https://github.com/3scale/porta/blob/512e40e5528ed217f84d373fc98e05a4c164e9a4/app/mailers/provider_invitation_mailer.rb#L13, and getting the email message rendered. Anyway, what I've just described does not belong to this test file because it's not an attribution of the worker; it's either an integration test or part of mailer's test.

On the other hand, if all you wanted to do was testing that the worker invokes `ProviderInvitationMailer#invitation`, then rename the test and remove the CMS template because it's not needed.

Reply:
OK I didn't know it was being stubbed.